components/imageproc/src/lib.rs

```diff
@@ -0,0 +1,326 @@
+        hasher.write_u8(u8::from(*self));
+        if let Some(w) = self.width()  { hasher.write_u32(w); }
+        if let Some(h) = self.height() { hasher.write_u32(h); }
+    }
+}
+
+/// Holds all data needed to perform a resize operation
+#[derive(Debug, PartialEq, Eq)]
+pub struct ImageOp {
+    source: String,
+    op: ResizeOp,
+    quality: u8,
+    hash: u64,
+    collision: Option<u32>,
+}
+
+impl ImageOp {
+    pub fn new(source: String, op: ResizeOp, quality: u8) -> ImageOp {
+        let mut hasher = XxHash::with_seed(0);
+        hasher.write(source.as_ref());
+        op.hash(&mut hasher);
+        hasher.write_u8(quality);
+        let hash = hasher.finish();
+
+        ImageOp { source, op, quality, hash, collision: None }
+    }
+
+    pub fn from_args(source: String, op: &str, width: Option<u32>, height: Option<u32>, quality: u8) -> Result<ImageOp> {
+        let op = ResizeOp::from_args(op, width, height)?;
+        Ok(Self::new(source, op, quality))
+    }
+
+    fn num_colli(&self) -> u32 { self.collision.unwrap_or(0) }
+
+    fn perform(&self, content_path: &Path, target_path: &Path) -> Result<()> {
+        use ResizeOp::*;
+
+        let src_path = content_path.join(&self.source);
+        if !ufs::file_stale(&src_path, target_path) {
+            return Ok(())
+        }
+
+        let mut img = image::open(&src_path)?;
+        let (img_w, img_h) = img.dimensions();
+
+        const RESIZE_FILTER: FilterType = FilterType::Gaussian;
+        const RATIO_EPSILLION: f32 = 0.1;
+
+        let img = match self.op {
+            Scale(w, h) => img.resize_exact(w, h, RESIZE_FILTER),
+            FitWidth(w) => img.resize(w, u32::max_value(), RESIZE_FILTER),
+            FitHeight(h) => img.resize(u32::max_value(), h, RESIZE_FILTER),
+            Fit(w, h) => img.resize(w, h, RESIZE_FILTER),
+            Fill(w, h) => {
+                let fw = img_w as f32 / w as f32;
+                let fh = img_h as f32 / h as f32;
+
+                if (fw - fh).abs() <= RATIO_EPSILLION {
+                    // The aspect is similar enough that there's not much point in cropping
+                    img.resize_exact(w, h, RESIZE_FILTER)
+                } else {
+                    // We perform the fill such that a crop is performed first and then resize_exact can be used,
+                    // which should be cheaper than resizing and then cropping (smaller number of pixels to resize).
+                    let (crop_w, crop_h) = match fw < fh {
+                        true  => (img_w, (fw * h as f32).round() as u32),
+                        false => ((fh * w as f32).round() as u32, img_h),
+                    };
+                    let (off_w, off_h) = match fw < fh {
+                        true  => (0, (img_h - crop_h) / 2),
+                        false => ((img_w - crop_w) / 2, 0),
+                    };
+                    img.crop(off_w, off_h, crop_w, crop_h).resize_exact(w, h, RESIZE_FILTER)
```

Comment:
`w`, `h`, `crop_w`, `crop_h` are pretty self-explanatory but can you lenghten fw, fh (I'm guessing `factor_w`?) and off_w, off_h (`offset_w`?)

Reply:
Yes, `f` is for `factor`, I'll update...